File: iris-ld/iris-ld-proofs/src/main/java/city/smartb/iris/ldproof/signer/LdProofSigner.java
package city.smartb.iris.ldproof.signer;

import java.security.GeneralSecurityException;
import java.util.Collections;

import com.nimbusds.jose.JOSEException;
import com.nimbusds.jose.JWSHeader;
import com.nimbusds.jose.JWSSigner;
import com.nimbusds.jose.util.Base64URL;

import city.smartb.iris.crypto.rsa.signer.Signer;
import city.smartb.iris.ldproof.LdJsonObjectBuilder;
import city.smartb.iris.ldproof.LdProof;
import city.smartb.iris.ldproof.LdProofBuilder;
import city.smartb.iris.ldproof.util.JWSUtil;
import city.smartb.iris.ldproof.util.SHAUtil;

public abstract class LdProofSigner {

	private final LdProofBuilder ldProofBuilder;

	protected final String signatureType;
	protected final Signer signer;

	protected LdProofSigner(String signatureType, Signer signer, LdProofBuilder ldProofBuilder) {
		this.signatureType = signatureType;
		this.signer = signer;
		this.ldProofBuilder = ldProofBuilder;
	}

	public LdProof sign(LdJsonObjectBuilder jsonLdObject) throws GeneralSecurityException {
		String canonicalizedDocument = jsonLdObject.buildCanonicalizedDocument();
		String canonicalizedProofOptions = ldProofBuilder.canonicalize(signer);
		String jws = sign(canonicalizedDocument, canonicalizedProofOptions);
		return ldProofBuilder.build(jws);
	}

	private String sign(String canonicalizedDocument, String canonicalizedProofOptions) throws GeneralSecurityException {
		byte[] signingInput = new byte[64];
		System.arraycopy(SHAUtil.sha256(canonicalizedProofOptions), 0, signingInput, 0, 32);
		System.arraycopy(SHAUtil.sha256(canonicalizedDocument), 0, signingInput, 32, 32);
		return this.sign(signingInput);
	}

	public String sign(byte[] signingInput) throws GeneralSecurityException {
		try {

			JWSHeader jwsHeader = new JWSHeader.Builder(signer.getAlgorithm()).base64URLEncodePayload(false).criticalParams(Collections.singleton("b64")).build();
			byte[] jwsSigningInput = JWSUtil.getJwsSigningInput(jwsHeader, signingInput);

			JWSSigner jwsSigner = new JoseJWSSigner(signer, signer.getAlgorithm());
			Base64URL signature = jwsSigner.sign(jwsHeader, jwsSigningInput);
			return JWSUtil.serializeDetachedJws(jwsHeader, signature);

		} catch (JOSEException ex) {
			throw new GeneralSecurityException("Error signing ", ex);
		}
	}

}

File: iris-ld/iris-ld-proofs/src/main/kotlin/city/smartb/iris/ld/ldproof/signer/LdProofSigner.kt
package city.smartb.iris.ld.ldproof.signer

import city.smartb.iris.crypto.dsl.signer.Signer
import city.smartb.iris.ld.ldproof.LdProof
import city.smartb.iris.ld.ldproof.LdProofBuilder
import city.smartb.iris.ld.ldproof.VerifiableJsonLdBuilder
import city.smartb.iris.ld.ldproof.util.JWSUtil
import city.smartb.iris.ld.ldproof.util.SHAUtil
import com.nimbusds.jose.JOSEException
import com.nimbusds.jose.JWSHeader
import com.nimbusds.jose.JWSSigner
import java.security.GeneralSecurityException

abstract class LdProofSigner protected constructor(
    protected val signatureType: String,
    protected val signer: Signer,
    private val ldProofBuilder: LdProofBuilder
) {
    @Throws(GeneralSecurityException::class)
    fun sign(jsonLdObject: VerifiableJsonLdBuilder): LdProof {
        val canonicalizedDocument = jsonLdObject.buildCanonicalizedDocument()
        val canonicalizedProofOptions = ldProofBuilder.canonicalize(signer)
        val jws = sign(canonicalizedDocument, canonicalizedProofOptions)
        return ldProofBuilder.build(jws)
    }

    @Throws(GeneralSecurityException::class)
    private fun sign(canonicalizedDocument: String, canonicalizedProofOptions: String): String {
        val signingInput = ByteArray(64)
        System.arraycopy(SHAUtil.sha256(canonicalizedProofOptions), 0, signingInput, 0, 32)
        System.arraycopy(SHAUtil.sha256(canonicalizedDocument), 0, signingInput, 32, 32)
        return this.sign(signingInput)
    }

    @Throws(GeneralSecurityException::class)
    fun sign(signingInput: ByteArray): String {
        return try {
            val jwsHeader =
                JWSHeader.Builder(signer.algorithm).base64URLEncodePayload(false).criticalParams(setOf("b64")).build()
            val jwsSigningInput = JWSUtil.getJwsSigningInput(jwsHeader, signingInput)
            val jwsSigner: JWSSigner = JoseJWSSigner(signer, signer.algorithm)
            val signature = jwsSigner.sign(jwsHeader, jwsSigningInput)
            JWSUtil.serializeDetachedJws(jwsHeader, signature)
        } catch (ex: JOSEException) {
            throw GeneralSecurityException("Error signing ", ex)
        }
    }
}
